hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/KafkaOffsetGen.java

```diff
@@ -107,48 +112,119 @@ public static String offsetsToStr(OffsetRange[] ranges) {
+          if (needSplitToMinPartitions) {
+            remainingPartitions = minPartitions - finalRanges.size();
+          }
+          long eventsPerPartition = (long) Math.ceil((1.0 * remainingEvents) / remainingPartitions);
+
           OffsetRange range = ranges[i];
-          if (!exhaustedPartitions.contains(range.partition())) {
-            long toOffsetMax = toOffsetMap.get(range.topicPartition());
-            long toOffset = Math.min(toOffsetMax, range.untilOffset() + eventsPerPartition);
-            if (toOffset == toOffsetMax) {
-              exhaustedPartitions.add(range.partition());
-            }
-            allocedEvents += toOffset - range.untilOffset();
-            // We need recompute toOffset if allocedEvents larger than numEvents.
-            if (allocedEvents > numEvents) {
-              long offsetsToAdd = Math.min(eventsPerPartition, (numEvents - allocedEvents));
-              toOffset = Math.min(toOffsetMax, toOffset + offsetsToAdd);
-            }
-            ranges[i] = OffsetRange.create(range.topicPartition(), range.fromOffset(), toOffset);
+          if (exhaustedPartitions.contains(range.partition())) {
+            continue;
           }
+
+          long toOffset = Math.min(range.untilOffset(), range.fromOffset() + eventsPerPartition);
+          if (toOffset == range.untilOffset()) {
+            exhaustedPartitions.add(range.partition());
+          }
+          allocedEvents += toOffset - range.fromOffset();
+          // We need recompute toOffset if allocedEvents larger than numEvents.
+          if (allocedEvents > numEvents) {
+            long offsetsToAdd = Math.min(eventsPerPartition, (numEvents - allocedEvents));
+            toOffset = Math.min(range.untilOffset(), toOffset + offsetsToAdd);
+          }
+          OffsetRange thisRange = OffsetRange.create(range.topicPartition(), range.fromOffset(), toOffset);
+          // filter out the empty ranges
+          if (thisRange.count() > 0) {
+            finalRanges.add(thisRange);
+          }
+          ranges[i] = OffsetRange.create(range.topicPartition(), range.fromOffset() + thisRange.count(), range.untilOffset());
+          allocatedPartitionsThisLoop.add(range.partition());
         }
       }
 
-      return ranges;
+      if (!needSplitToMinPartitions) {
+        LOG.debug("final ranges merged by topic partition {}", Arrays.toString(mergeRangesByTopicPartition(finalRanges.toArray(new OffsetRange[0]))));
+        return mergeRangesByTopicPartition(finalRanges.toArray(new OffsetRange[0]));
+      }
+      finalRanges.sort(SORT_BY_PARTITION);
+      LOG.debug("final ranges {}", Arrays.toString(finalRanges.toArray(new OffsetRange[0])));
+      return finalRanges.toArray(new OffsetRange[0]);
+    }
+
+    // the number of the result ranges can be less or more than minPartitions due to rounding
+    public static OffsetRange[] splitRangesToMinPartitions(OffsetRange[] oldRanges, long minPartitions) {
```

Comment:
Looks like this method is not being used. Can we remove it ?

Reply:
sure